### PR TITLE
[mysql] Streaming merge during snapshotSplit read (#1219)

### PR DESCRIPTION
**Pull request by fuyun2024**
issues :  #1219 
This PR supports streaming merge during snapshot split reads

Changes:

1、Change snapshot split queue size = splitSize + 1024
2、Once a batch of data is obtained, it can be merged immediately without waiting for all data to be collected

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 1]
```diff
@@ -229,23 +240,54 @@ public Iterator<SourceRecord> pollSplitRecords() throws InterruptedException {
...
+                    SourceRecord record = event.getRecord();
+                    if (lowWatermark == null) {
+                        lowWatermark = record;
+                        checkLowWatermark(lowWatermark);
+                        continue;
+                    }
+
+                    if (highWatermark == null && isHighWatermarkEvent(record)) {
+                        highWatermark = record;
+                        checkHighWatermark(highWatermark);
```
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:261]: I think `checkHighWatermark` should be like this : 
```
private void checkHighWatermark(SourceRecord highWatermark) {
        checkState(
                highWatermark != null && isHighWatermarkEvent(highWatermark),
                "The last record should be high watermark signal event, but is " + highWatermark);
}
```
And `checkHighWatermark` should be invoked after the whole loop.
```
while (!reachBinlogEnd) {
   ......
}
checkHighWatermark(highWatermark);
```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 2]
```diff
@@ -262,6 +304,36 @@ private void checkReadException() {
...
+
+    private boolean checkValidBinlogRecord(SourceRecord record) {
+        if (isDataChangeRecord(record)) {
+            Object[] key =
+                    getSplitKey(currentSnapshotSplit.getSplitKeyType(), record, nameAdjuster);
+            if (splitKeyRangeContains(
+                    key,
+                    currentSnapshotSplit.getSplitStart(),
+                    currentSnapshotSplit.getSplitEnd())) {
+                return true;
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:331]:
```
return splitKeyRangeContains(
                    key,
                    currentSnapshotSplit.getSplitStart(),
                    currentSnapshotSplit.getSplitEnd());
```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java [hunk 3]
```diff
@@ -75,125 +73,50 @@ public static Object[] rowToArray(ResultSet rs, int size) throws SQLException {
...
-            SourceRecord highWatermark = null;
-            int i = 1;
-            for (; i < sourceRecords.size(); i++) {
-                SourceRecord sourceRecord = sourceRecords.get(i);
-                if (!isHighWatermarkEvent(sourceRecord)) {
-                    snapshotRecords.put((Struct) sourceRecord.key(), sourceRecord);
-                } else {
-                    highWatermark = sourceRecord;
-                    i++;
+    /** upsert binlog events to snapshot events collect. */
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/source/utils/RecordUtils.java:76]: collect -> collection

**Review by ruanhang1993 — COMMENTED** [review]
I have read this PR. And I think there are some parts that need to be modified.

**Comment by fuyun2024** [comment]
Hi @ruanhang1993 thanks for your advice. I submitted the code again.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java]: Thanks for your work, @fuyun2024 . I think we still need the validation method about the high watermark after this loop. In the new code, you changed to this logic:
```
while (!reachBinlogEnd) {
    ......
    if (reachBinlogStart && RecordUtils.isEndWatermarkEvent(record)) {
        // capture to end watermark events, stop the loop
        reachBinlogEnd = true;
        break;
    }
    ......
}
```
I am afraid that if the high watermark does not appear, this loop will not be able to end. The `reachBinlogStart` should not be added in the `if` code block.

So I think we still need the validation about the high watermark, and move this validation to the place like I mentioned above. What do you think?

[the remark this answers is not in the document]
Comment (ruanhang1993) [reply flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java]: IMO, I prefer to obey the original logic, which will end by receiving the binglog-end event, no matter whether the high watermark occurs. I am not sure if there is a situation that the high watermark disappear. Maybe other reviewers could help to confirm this change.

**Review by leonardBang — APPROVED** [review]
Thanks @fuyun2024 for the contribution, LGTM

**Comment by leonardBang** [comment]
Thanks @fuyun2024 and @ruanhang1993 for the great work, I think this feature will help users a lot

**Pull request merged**